core/arch/interp.c

```diff
@@ -3908,7 +3908,7 @@ build_bb_ilist(dcontext_t *dcontext, build_bb_t *bb)
             BBPRINT(bb, 3, "reached end pc " PFX ", stopping\n", bb->stop_pc);
             break;
         }
-        if (total_instrs > DYNAMO_OPTION(max_bb_instrs)) {
+        if (total_instrs == DYNAMO_OPTION(max_bb_instrs) - 1) {
```

Comment:
So this is a swing of 2: but I think we only need to reduce by 1?  The artificial jump added on truncation is not passed to clients: it's like other mangling added later.  So if `-max_bb_instrs 4` means 4 app instrs passed to clients, we'd want ==, not ==-1, right?

Plus, if we did want -1 and to stop at 3 instead of 4, this is not sufficient: the inner loop above currently has `total_instrs <= cur_max_bb_instrs`.

Reply:
I see a bug in a prior change: this should use `cur_max_bb_instrs` to handle single-step, rather than reading the option.

Reply:
Should we add a sanity check at checklevel 1 or sthg (`DOCHECK(1, ...)`) in client_process_bb() that counts the app instrs and ensures the count is <= DYNAMO_OPTION(max_bb_instrs)?

Reply:
> 
> 
> So this is a swing of 2: but I think we only need to reduce by 1? The artificial jump added on truncation is not passed to clients: it's like other mangling added later. So if `-max_bb_instrs 4` means 4 app instrs passed to clients, we'd want ==, not ==-1, right?
> 
> Plus, if we did want -1 and to stop at 3 instead of 4, this is not sufficient: the inner loop above currently has `total_instrs <= cur_max_bb_instrs`.

With the default (-max_bb_instrs 256), the block size was going to 257 but OPND_CREATE_INT8 can only handle 255. I now realize that 256 is not a hard limit and the user could specify a higher value. In that case, is it better to replace the OPND_CREATE_INT8 with OPND_CREATE_INT16?

Reply:
> is it better to replace the OPND_CREATE_INT8 with OPND_CREATE_INT16?

Yes I agree that should be done since IIRC the encoder will still use the 1-byte immed if it fits.